encoding/time_encode.go

```diff
@@ -1,24 +1,146 @@
 package encoding
 
 import (
-	"errors"
+	"fmt"
 	"io"
+	"math"
+	"math/bits"
+	"time"
+
+	bitstream "github.com/dgryski/go-bitstream"
+	"github.com/xichen2020/eventdb/generated/proto/encodingpb"
+	"github.com/xichen2020/eventdb/x/proto"
 )
 
 // TimeEncoder encodes times values.
 type TimeEncoder interface {
 	// Encode encodes a collection of time values and writes the encoded bytes to the writer.
 	// Callers should explicitly call `Reset` before subsequent call to `Encode`.
-	Encode(writer io.Writer, values ForwardTimeIterator) error
+	Encode(writer io.Writer, values RewindableTimeIterator) error
 }
 
 // TimeEnc is a int encoder.
-type TimeEnc struct{}
+type TimeEnc struct {
+	metaProto encodingpb.TimeMeta
+	bitWriter *bitstream.BitWriter
+	buf       []byte
+}
 
 // NewTimeEncoder creates a new time encoder.
-func NewTimeEncoder(writer io.Writer) *TimeEnc { return &TimeEnc{} }
+func NewTimeEncoder() *TimeEnc {
+	return &TimeEnc{
+		bitWriter: bitstream.NewWriter(nil),
+	}
+}
 
 // Encode encodes a collection of time values and writes the encoded bytes to the writer.
-func (enc *TimeEnc) Encode(writer io.Writer, values ForwardTimeIterator) error {
-	return errors.New("not implemented")
+func (enc *TimeEnc) Encode(
+	writer io.Writer,
+	valuesIt RewindableTimeIterator,
+	resolution time.Duration,
```

Comment:
Can we define a struct `EncodeTimeOptions` for this? I'd imagine we need similar constructs for block encoding too.
```
type EncodeTimeOptions struct {
  Resolution time.Duration
}

func (enc *TimeEnc) Encode(writer io.Writer, valuesIt RewindableTimeIterator, opts EncodeTimeOptions) {
  ...
}
```